diff -uNr avahi-0.6.31/avahi-daemon/static-services.c new-avahi-0.6.31/avahi-daemon/static-services.c
--- avahi-0.6.31/avahi-daemon/static-services.c	2010-08-26 08:51:39.002153001 +0800
+++ new-avahi-0.6.31/avahi-daemon/static-services.c	2015-12-16 16:45:18.043704394 +0800
@@ -49,6 +49,9 @@
 #include "main.h"
 #include "static-services.h"
 
+#undef AVAHI_SERVICE_DIR
+#define AVAHI_SERVICE_DIR "/tmp/avahi_service/services"
+
 typedef struct StaticService StaticService;
 typedef struct StaticServiceGroup StaticServiceGroup;
 
@@ -697,8 +700,7 @@
 
     memset(&globbuf, 0, sizeof(globbuf));
 
-    if ((globret = glob(in_chroot ? "/services/*.service" : AVAHI_SERVICE_DIR "/*.service", GLOB_ERR, NULL, &globbuf)) != 0)
-
+    if ((globret = glob(in_chroot ? "/services/*.service" : AVAHI_SERVICE_DIR "/*.service", GLOB_ERR, NULL, &globbuf)) != 0) 
         switch (globret) {
 #ifdef GLOB_NOSPACE
 	    case GLOB_NOSPACE:
