diff -Nur avahi-0.6.31/avahi-daemon/main.c avahi-0.6.31-mod/avahi-daemon/main.c
--- avahi-0.6.31/avahi-daemon/main.c	2010-11-18 11:00:26.288465933 +0800
+++ avahi-0.6.31-mod/avahi-daemon/main.c	2014-04-14 23:42:14.000000000 +0800
@@ -381,6 +381,7 @@
             n = avahi_alternative_host_name(avahi_server_get_host_name(s));
 
             avahi_log_warn("Host name conflict, retrying with %s", n);
+            usleep(1000000);
             sd_notifyf(0, "STATUS=Host name conflict, retrying with %s", n);
             avahi_set_proc_title(argv0, "%s: collision [%s]", argv0, n);
 
diff -Nur avahi-0.6.31/avahi-daemon/static-services.c avahi-0.6.31-mod/avahi-daemon/static-services.c
--- avahi-0.6.31/avahi-daemon/static-services.c	2010-08-26 08:51:39.002153001 +0800
+++ avahi-0.6.31-mod/avahi-daemon/static-services.c	2014-04-03 13:46:12.000000000 +0800
@@ -227,8 +227,8 @@
     if (g->entry_group && !avahi_s_entry_group_is_empty(g->entry_group))
         /* This service group is already registered in the server */
         return;
-
-    if (!g->chosen_name || (g->replace_wildcards && strstr(g->name, "%h"))) {
+        
+    if (!g->chosen_name && (g->replace_wildcards && strstr(g->name, "%h"))) {
 
         avahi_free(g->chosen_name);
 
@@ -242,9 +242,12 @@
             g->chosen_name = replacestr(g->name, "%h", label);
         } else
             g->chosen_name = avahi_strdup(g->name);
-
+    }else{
+    	if(!g->chosen_name){
+			g->chosen_name = avahi_strdup(g->name);
+		}
     }
-
+   
     if (!g->entry_group)
         g->entry_group = avahi_s_entry_group_new(avahi_server, entry_group_callback, g);
 
