--- linux-kernel.orig/fs/fat/file.c	2017-12-03 09:20:46.899331195 +0800
+++ linux-kernel.mod/fs/fat/file.c	2017-12-03 09:26:46.168219939 +0800
@@ -318,7 +318,7 @@
 		MSDOS_I(inode)->i_logstart = 0;
 	}
 	MSDOS_I(inode)->i_attrs |= ATTR_ARCH;
-	inode->i_ctime = inode->i_mtime = CURRENT_TIME_SEC;
+
 	if (wait) {
 		err = fat_sync_inode(inode);
 		if (err) {
@@ -362,7 +362,7 @@
 		free_start = ret;
 	}
 	inode->i_blocks = skip << (MSDOS_SB(sb)->cluster_bits - 9);
-
+	inode->i_ctime = inode->i_mtime = CURRENT_TIME_SEC;
 	/* Freeing the remained cluster chain */
 	return fat_free_clusters(inode, free_start);
 }
